internal/resourcecache/resourcecache.go

```diff
@@ -0,0 +1,109 @@
+package resourcecache
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+var sleepTimeBeforeCleanup = 2 * time.Minute
+
+// ResourceCache is a structure that saves information about a nearly created resource.
+// Resources can be added and retrieved from the cache. A retrieval (ResourceFromCache) also removes the item from the cache.
+// If `sleepTimeBeforeCleanup` has passed, and the resource has not yet been retrieved, ResourceCache considers that resource stale,
+// and removes it.
+type ResourceCache struct {
+	resources map[string]*Resource
+	sync.Mutex
+}
+
+// Resource contains the actual resource itself (which must implement the IdentifiableCreatable interface),
+// as well as stores function pointers that pertain to how that resource should be cleaned up.
+type Resource struct {
+	resource     IdentifiableCreatable
+	cleanupFuncs []func()
+}
+
+// IdentifiableCreatable are the qualities needed by the caller of the resource.
+// Once a resource is retrieved, SetCreated() will be called, indicating to the server
+// that resource is ready to be listed and operated upon, and ID() will be used to identify the
+// newly created resource to the server.
+type IdentifiableCreatable interface {
+	ID() string
+	SetCreated()
+}
+
+// New creates a new ResourceCache
+func New() *ResourceCache {
+	rc := &ResourceCache{
+		resources: make(map[string]*Resource),
+	}
+	return rc
+}
+
+// AddResourceToCache takes a unique resource name (retrieved from the client request, not generated by the server)
+// a newly created resource, and functions to cleanup that newly created resource.
+// It adds the Resource to the ResourceCache, as well as starts a go routine that is responsible for cleaning up the
+// resource if the server has not gotten another request for it (i.e. if ResourceFromCache has not been called).
+// AddResourceToCache expects `name`s passed to it to be unique. If a duplicate `name` is detected, it returns an error.
+func (rc *ResourceCache) AddResourceToCache(name string, resource IdentifiableCreatable, cleanupFuncs []func()) error {
+	return rc.AddResourceToCacheWithTimeout(name, resource, cleanupFuncs, sleepTimeBeforeCleanup)
+}
+
+// AddResourceToCacheWithTimeout is used for testing purposes. It allows the caller to set the timeout, allowing for faster tests.
+// Most callers should use AddResourceToCache instead.
+func (rc *ResourceCache) AddResourceToCacheWithTimeout(name string, resource IdentifiableCreatable, cleanupFuncs []func(), timeout time.Duration) error {
+	rc.Lock()
+	defer rc.Unlock()
+
+	if _, ok := rc.resources[name]; ok {
+		return errors.Errorf("failed to add entry %s to ResourceCache; entry already exists", name)
+	}
+	rc.resources[name] = &Resource{
+		resource:     resource,
+		cleanupFuncs: cleanupFuncs,
+	}
+	go rc.cleanupResourceAfterSleep(name, timeout)
```

Comment:
Depending on number of resources you want to store, a goroutine per resource might not be a very scalable model.

The alternative is a single goroutine which wakes up say once a minute (or 10 minutes) and cleans up all stale entries. If you need better precision, you can check if the object is stale upon retrieve.

Reply:
fair enough, I've moved to one goroutine. I can't imagine there would be that many, but why not future proof it.